File: core/sign_rt/io/keypoint_codec.py
"""Compact keypoint codec for the WebSocket wire format.

PORTED to TS — encode/decode must be byte-compatible (keypoint_codec.ts).

Wire format (little-endian):
    offset 0  : uint8  magic[0] = 'S' (0x53)
    offset 1  : uint8  magic[1] = 'R' (0x52)
    offset 2  : uint8  version  = 1
    offset 3  : uint8  ndim     = 3      (x, y, visibility)
    offset 4  : uint16 n_keypoints
    offset 6+ : int16  body, row-major (n_keypoints * ndim) values,
                each = round(clamp(value, -CLAMP, CLAMP) * SCALE)

Quantization resolution is 1/SCALE ≈ 2.4e-4 — lossless within tolerance for
shoulder-normalized coordinates. Privacy: payload carries keypoints only,
never pixels.
"""
from __future__ import annotations

import numpy as np

MAGIC0 = 0x53  # 'S'
MAGIC1 = 0x52  # 'R'
VERSION = 1
NDIM = 3
HEADER_BYTES = 6

SCALE = 4096.0
CLAMP = 8.0  # normalized coords + visibility comfortably fit in [-8, 8]
_INT16_MIN = -32768
_INT16_MAX = 32767


def encode(kpts: np.ndarray) -> bytes:
    """Pack a (N_KEYPOINTS, 3) float array into a compact byte payload.

    Privacy: payload carries keypoints only, never pixels.
    """
    arr = np.asarray(kpts, dtype=np.float32)
    if arr.ndim != 2 or arr.shape[1] != NDIM:
        raise ValueError(f"expected (N, {NDIM}) keypoints, got {arr.shape}")
    n = arr.shape[0]

    flat = np.nan_to_num(arr, nan=0.0, posinf=CLAMP, neginf=-CLAMP).reshape(-1)
    q = np.clip(np.rint(np.clip(flat, -CLAMP, CLAMP) * SCALE), _INT16_MIN, _INT16_MAX)
    q = q.astype("<i2")

    header = bytes([MAGIC0, MAGIC1, VERSION, NDIM]) + int(n).to_bytes(2, "little")
    return header + q.tobytes()


def decode(payload: bytes) -> np.ndarray:
    """Inverse of encode(): bytes -> (N_KEYPOINTS, 3) float array."""
    if len(payload) < HEADER_BYTES:
        raise ValueError("payload too short for header")
    if payload[0] != MAGIC0 or payload[1] != MAGIC1:
        raise ValueError("bad magic")
    if payload[2] != VERSION:
        raise ValueError(f"unsupported version {payload[2]}")
    ndim = payload[3]
    n = int.from_bytes(payload[4:6], "little")

    body = np.frombuffer(payload, dtype="<i2", offset=HEADER_BYTES, count=n * ndim)
    return (body.astype(np.float32) / SCALE).reshape(n, ndim)

File: frontend/src/core_port/keypoint_codec.ts
// Ported from core/sign_rt/io/keypoint_codec.py.
// MUST be byte-compatible with the Python encoder/decoder (wire format).
//
// Layout (little-endian):
//   0: uint8  'S' (0x53)   1: uint8  'R' (0x52)
//   2: uint8  version=1    3: uint8  ndim=3
//   4: uint16 n_keypoints
//   6: int16  body, row-major (n * ndim), each = round(clamp(v,-8,8) * 4096)

export const MAGIC0 = 0x53;
export const MAGIC1 = 0x52;
export const VERSION = 1;
export const NDIM = 3;
export const HEADER_BYTES = 6;

export const SCALE = 4096.0;
export const CLAMP = 8.0;
const INT16_MIN = -32768;
const INT16_MAX = 32767;

function clamp(v: number, lo: number, hi: number): number {
  return v < lo ? lo : v > hi ? hi : v;
}

export function encode(kpts: Float32Array): ArrayBuffer {
  const total = kpts.length;
  if (total % NDIM !== 0) throw new Error(`expected multiple of ${NDIM}, got ${total}`);
  const n = total / NDIM;

  const buf = new ArrayBuffer(HEADER_BYTES + n * NDIM * 2);
  const view = new DataView(buf);
  view.setUint8(0, MAGIC0);
  view.setUint8(1, MAGIC1);
  view.setUint8(2, VERSION);
  view.setUint8(3, NDIM);
  view.setUint16(4, n, true);

  for (let i = 0; i < total; i++) {
    let v = kpts[i];
    if (!Number.isFinite(v)) v = 0;
    // Math.round matches numpy rint for the .5-free values quantization produces
    // in practice; use round-half-away to stay deterministic across surfaces.
    const q = clamp(Math.round(clamp(v, -CLAMP, CLAMP) * SCALE), INT16_MIN, INT16_MAX);
    view.setInt16(HEADER_BYTES + i * 2, q, true);
  }
  return buf;
}

export function decode(payload: ArrayBuffer): Float32Array {
  if (payload.byteLength < HEADER_BYTES) throw new Error("payload too short");
  const view = new DataView(payload);
  if (view.getUint8(0) !== MAGIC0 || view.getUint8(1) !== MAGIC1) throw new Error("bad magic");
  if (view.getUint8(2) !== VERSION) throw new Error(`unsupported version ${view.getUint8(2)}`);
  const ndim = view.getUint8(3);
  const n = view.getUint16(4, true);

  const out = new Float32Array(n * ndim);
  for (let i = 0; i < n * ndim; i++) {
    out[i] = view.getInt16(HEADER_BYTES + i * 2, true) / SCALE;
  }
  return out;
}
